src/webui/webapplication.cpp

```diff
@@ -700,11 +710,46 @@ bool WebApplication::validateHostHeader(const QStringList &domains) const
     }
 
     LogMsg(tr("WebUI: Invalid Host header. Request source IP: '%1'. Received Host header: '%2'")
-           .arg(m_env.clientAddress.toString(), m_request.headers[Http::HEADER_HOST])
+           .arg(m_clientAddress.toString(), m_request.headers[Http::HEADER_HOST])
             , Log::WARNING);
     return false;
 }
 
+QHostAddress WebApplication::resolveClientAddress() const
+{
+    QHostAddress clientAddress = m_env.clientAddress;
+
+    if (!m_isReverseProxyEnabled)
+        return clientAddress;
+
+    // Only reverse proxy can overwrite client address
+    if (clientAddress.isEqual(m_reverseProxyAddress))
+    {
+        const QString forwardedFor = m_request.headers.value(Http::HEADER_X_FORWARDED_FOR);
+
+        if (!forwardedFor.isEmpty())
+        {
+            // client address is the 1st global IP in X-Forwarded-For or, if none available, the 1st IP in the list
```

Comment:
IIRC, client address is always first. Any subsequent addresses become to midway proxies. Am I wrong?

Reply:
Yes, you are wrong but it is a fairly common error.

When I access to qbittorrent from my work, I have the following the X-forwarded-for 
172.29.225.237,194.254.4.90,10.0.3.254,127.0.0.1

* First IP is the local IP of my desktop PC at work.
* 2nd IP is the public IP used by my work proxy.
* 3rd IP is the local IP of my reverse proxy at home. This reverse proxy runs on my gateway and performs SSL termination.
* 4th IP is added by nginx running on the same computer as qbittorrent (proxy_pass).

The correct IP is always the last public IP, here, the 2nd IP.

Reply:
>Yes, you are wrong but it is a fairly common error.

I was only guided by https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For#syntax

Reply:
It is correct, the first IP is the real client IP but it does not mean it is a public IP, that's why it is kinda misleading.

The worst case I ever seen is something like that (combination of proxy, VPN and reverse proxies):
192.168.0.1, public_IP, 10.0.23.1, public_IP2, 172.29.226.1, 127.0.0.1

In this case, the correct public IP is public_IP2 (last public IP). I admit, it is a very unusual settings but it is a valid value.

Reply:
OK, thanks.

One more question.
We specify the address of the proxy server to make sure that the request (containing `X-Forwarded-For` header) is received from it. This assumes that requests can come from both the proxy server address and another address. Is it a valid usecase to have a configured proxy server and at the same time some other access channels for untrusted clients?

Reply:
> The correct IP is always the last public IP, here, the 2nd IP.

That's interesting what's your proxy configuration ? Using HAProxy I only have the public IP in the header.
```
backend bk_qbittorrent
  http-response set-header Strict-Transport-Security "max-age=16000000; includeSubDomains; preload;"
  http-response set-header X-Frame-Options "DENY"
  http-response set-header X-XSS-Protection "1;mode=block"
  http-response set-header X-Content-Type-Options "nosniff"
  http-response set-header Referrer-Policy "same-origin"
  option forwardfor
  server server1 127.0.0.1:8040 check
```

Log from here:
```
 if (!forwardedFor.isEmpty())
        {
            LogMsg(tr("FORWARDED_FOR value:%1").arg(forwardedFor));
            // client address is the 1st global IP in X-Forwarded-For or, if none available, the 1st IP in the list
            const QStringList remoteIpList = forwardedFor.split(",");
```

Reply:
> Is it a valid usecase to have a configured proxy server and at the same time some other access channels for untrusted clients?

Actually I'm using it in this case. I run HAProxy for remote connections, locally I checked "Bypass authentication for clients on localhost" and use a python script to clear old torrents.

Reply:
> That's interesting what's your proxy configuration ? Using HAProxy I only have the public IP in the header.

When connecting from work:
  local browser at work 
 => work proxy 
 => haproxy on personal gateway (SSL termination) 
 => haproxy on personal gateway (multi-site routing)
 => qbittorrent

I don't know why but it is not possible to have SSL termination + multi-site routing at the same time, haproxy has to send request to itself.

> Actually I'm using it in this case. I run HAProxy for remote connections, locally I checked "Bypass authentication for clients on localhost" and use a python script to clear old torrents.

So do I. I have haproxy running on my opnsense gateway and performing SSL termination for several sites. 

> We specify the address of the proxy server to make sure that the request (containing X-Forwarded-For header) is received from it. This assumes that requests can come from both the proxy server address and another address. Is it a valid usecase to have a configured proxy server and at the same time some other access channels for untrusted clients?

Technically, I don't use it that way as I use nginx authentication and reverse proxy but when I made my first patch I assumed someone might want to have both authentification methods enabled (webserver + qbittorrent) .

Reply:
I think for a safe reverse proxy configuration, it's better to delete `X-Forwarded-For` at frontend, so someone from the
outside cannot spoof this header. It also ensures the first IP in the header is the client IP.

Reply:
> I think for a safe reverse proxy configuration, it's better to delete X-Forwarded-For at frontend, so someone from the
outside cannot spoof this header. It also ensures the first IP in the header is the client IP.

I don't think it will change anything as the frontend will always add the incoming IP address at the end of X-Forwarded-For string. Anything existing before should be assumed unsafe and purely informational. That's why taking the last public IP is important.

Moreover, it is not always possible to reconfigure the reverse proxy.

Reply:
>We specify the address of the proxy server to make sure that the request (containing `X-Forwarded-For` header) is received from it. This assumes that requests can come from both the proxy server address and another address. Is it a valid usecase to have a configured proxy server and at the same time some other access channels for untrusted clients?

The point is that if it is only available through a proxy (as it should be, IMO; i.e., if we have a proxy, then it should only be available through it), then we do not need to have a proxy address check, but just specify whether it should resolve the client address or not.

It looks like we should still allow user to explicitly set the proxy address for better security. But maybe we can be more lenient in another case, i.e. allow the user to enable client address resolution without explicitly specifying the proxy address?

Reply:
> But maybe we can be more lenient in another case, i.e. allow the user to enable client address resolution without explicitly specifying the proxy address?

It's hard to say, people do it differently.
For example Gitea allows client address resolution from 127.0.0.0 by default, SFTPGo needs to specify proxy address whitelist.

Reply:
It's just that we could provide more options to the user by placing more responsibility on them (of course, by warning them about this). If he is confident in the security of his network environment settings, then allow him not to explicitly specify the proxy server address.

In any case, this is not a blocking comment. So if you don't agree, you can continue as is and wait for user feedback.